File: plugins/LayoutController/src/main/java/com/aliucord/plugins/layoutcontroller/patchers/base/BasePatcher.java
package com.aliucord.plugins.layoutcontroller.patchers.base;

import com.aliucord.api.PatcherAPI;
import com.aliucord.api.SettingsAPI;
import com.aliucord.patcher.PinePatchFn;
import com.aliucord.plugins.layoutcontroller.util.Const;

import java.lang.reflect.Member;

import top.canyie.pine.Pine;

public abstract class BasePatcher {

    public final String key;
    public final String description;

    private final Member classMember;

    public BasePatcher(String key, String description, Member classMember) {
        this.key = key;
        this.description = description;
        this.classMember = classMember;
    }

    public void patch(PatcherAPI patcher, SettingsAPI sets) {
        patcher.patch(classMember, new PinePatchFn(callFrame -> {
            if (sets.getBool(key, Const.PREFERENCE_DEFAULT_VALUE)) {
                try {
                    patchBody(callFrame);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            } else {
                callFrame.setResult(callFrame.getResult());
            }
        }));
    }

    public abstract void patchBody(Pine.CallFrame callFrame) throws Exception;

}

File: plugins/LayoutController/src/main/java/com/aliucord/plugins/layoutcontroller/patchers/base/BasePatcher.kt
package com.aliucord.plugins.layoutcontroller.patchers.base

import com.aliucord.api.PatcherAPI
import com.aliucord.api.SettingsAPI
import com.aliucord.patcher.Hook
import com.aliucord.plugins.layoutcontroller.util.PREFERENCE_DEFAULT_VALUE
import de.robv.android.xposed.XC_MethodHook
import java.lang.reflect.Member

abstract class BasePatcher(
    val key: String, val description: String, private val classMember: Member
) {

    fun patch(patcher: PatcherAPI, sets: SettingsAPI) {
        patcher.patch(classMember, Hook { callFrame ->
            if (sets.getBool(key, PREFERENCE_DEFAULT_VALUE)) {
                patchBody(callFrame)
            } else {
                callFrame.result = callFrame.result
            }
        })
    }

    abstract fun patchBody(callFrame: XC_MethodHook.MethodHookParam)
}